server/sonar-server/src/main/java/org/sonar/server/issue/index/IssueIndexer.java

```diff
@@ -70,51 +90,154 @@ public AuthorizationScope getAuthorizationScope() {
   }
 
-  public void index(Collection<String> issueKeys) {
-    doIndex(createBulkIndexer(Size.REGULAR), issueKeys);
+  @Override
+  public IndexingResult index(DbSession dbSession, Collection<EsQueueDto> items) {
+    ListMultimap<String, EsQueueDto> itemsByIssueKey = ArrayListMultimap.create();
+    ListMultimap<String, EsQueueDto> itemsByProjectKey = ArrayListMultimap.create();
+    items.forEach(i -> {
+      if (ID_TYPE_ISSUE_KEY.equals(i.getDocIdType())) {
+        itemsByIssueKey.put(i.getDocId(), i);
+      } else if (ID_TYPE_PROJECT_UUID.equals(i.getDocIdType())) {
+        itemsByProjectKey.put(i.getDocId(), i);
+      } else {
+        LOGGER.error("Unsupported es_queue.doc_id_type for issues. Manual fix is required: " + i);
+      }
+    });
+
+    IndexingResult result = new IndexingResult();
+    result.add(doIndexIssueItems(dbSession, itemsByIssueKey));
+    result.add(doIndexProjectItems(dbSession, itemsByProjectKey));
+    return result;
   }
 
-  private void doIndex(BulkIndexer bulk, Collection<String> issueKeys) {
-    try (IssueIterator issues = issueIteratorFactory.createForIssueKeys(issueKeys)) {
-      doIndex(bulk, issues);
+  private IndexingResult doIndexIssueItems(DbSession dbSession, ListMultimap<String, EsQueueDto> itemsByIssueKey) {
+    if (itemsByIssueKey.isEmpty()) {
+      return new IndexingResult();
+    }
+    IndexingListener listener = new OneToOneResilientIndexingListener(dbClient, dbSession, itemsByIssueKey.values());
+    BulkIndexer bulkIndexer = createBulkIndexer(Size.REGULAR, listener);
+    bulkIndexer.start();
+
+    try (IssueIterator issues = issueIteratorFactory.createForIssueKeys(itemsByIssueKey.keySet())) {
+      while (issues.hasNext()) {
+        IssueDoc issue = issues.next();
+        bulkIndexer.add(newIndexRequest(issue));
+        itemsByIssueKey.removeAll(issue.getId());
+      }
     }
+
+    // the remaining uuids reference issues that don't exist in db. They must
+    // be deleted from index.
+    itemsByIssueKey.values().forEach(
+      item -> bulkIndexer.addDeletion(INDEX_TYPE_ISSUE, item.getDocId(), item.getDocRouting()));
+
+    return bulkIndexer.stop();
   }
 
-  private void doIndex(BulkIndexer bulk, @Nullable String projectUuid) {
-    try (IssueIterator issues = issueIteratorFactory.createForProject(projectUuid)) {
-      doIndex(bulk, issues);
+  private IndexingResult doIndexProjectItems(DbSession dbSession, ListMultimap<String, EsQueueDto> itemsByProjectUuid) {
+    if (itemsByProjectUuid.isEmpty()) {
+      return new IndexingResult();
+    }
+
+    // one project, referenced by es_queue.doc_id = many issues
+    IndexingListener listener = new OneToManyResilientIndexingListener(dbClient, dbSession, itemsByProjectUuid.values());
+    BulkIndexer bulkIndexer = createBulkIndexer(Size.REGULAR, listener);
+    bulkIndexer.start();
+
+    for (String projectUuid : itemsByProjectUuid.keySet()) {
+      // TODO support loading of multiple projects in a single SQL request
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)